User-guide: Add Ironic installation in Metal3 (#252)
metal3-io/metal3-docs: docs/user-guide/src/bmo/ironic_installation.md

# Install Ironic

Baremetal Operator must be able to interact with Ironic. First,

> we will see what are the Ironic optional containers we need to
> run. After that we will go through Ironic installation in-cluster and
> out-of-cluster. In any case, there are a couple of containers that need
> to run in order to provision baremetal nodes.

Reviewer: ```suggestion
we will see what are the optional Ironic containers we need to
run. After that, we will go through Ironic installation options: in-cluster and
out-of-cluster. In any case, there are a couple of containers that need
to run in order to provision baremetal nodes.
```

- ironic
- ironic-inspector
- ironic-endpoint-keepalived
- ironic-log-watch
- ipa-downloader
- dnsmasq
- httpd

To know more about each container's functionality check the documentation
[here](https://github.com/metal3-io/ironic-image#description).

## Prerequisites

Container runtime (e.g., docker, podman). Here we use docker.

## Environmental variables

> <!-- include these inline so we can style an match variables -->

Reviewer: is this needed?

Author: no, removed

{{#include ironic_variables.md}}

## Optional containers

> In some circumstances, it might be necessary to install some additional containers on

Reviewer: ```suggestion
In some circumstances, it might be necessary to install some additional containers
```

> top of Ironic containers. And this might depend on whether your target nodes are

Reviewer 2: ```suggestion
In some circumstances, it might be necessary to install some additional containers besides the mandatory Ironic containers. And this might depend on whether your target nodes are
```

> besides the mandatory Ironic containers. And this might depend on whether your target nodes are

Reviewer: Can we make it clear and provide a list of mandatory and list of optional containers at the beginning since we mentioned it there as well, to avoid confusion about which container belongs to which list?

Author: I removed this section, we can add a section later on for development environment

virtual or physical machines, or whether you want to use more stable BMC protocol
like Redfish and IPMI. Let's check why we would need these optional containers.
We run these optional containers on the host machine, i.e., out-of-cluster regardless
of the Ironic installation method.

> ### httpd

Reviewer: Do we need this section? if yes, we might need to also include other container descriptions (dnsmasq, ironic-log-watch etc) as well.

Author: removed

We use this container to run an Apache web server for hosting node images via HTTP or via HTTPS when TLS enabled. Below is the list of images (from `$IMAGES_PATH`) that we mount into the httpd container's `/shared` path.

1. Ironic Python Images (IPA) via http for PXE booting
    - ironic-python-agent.kernel
    - ironic-python-agent.initramfs

2. Target node image in qcow2 format

>     - bionic-server-cloudimg-amd64-disk.img

Reviewer: you meant:
```suggestion
    - bionic-server-cloudimg-amd64-disk.qcow2
```
?

Author: no, either if fine for me

Reviewer: since you mentioned qcow format, it makes sense to note it accordingly

We are not directly using Apache container image here, but instead [quay.io/metal3-io/ironic](https://quay.io/repository/metal3-io/ironic?tag=latest&tab=tags) which has `runhttpd` entry point to start off the httpd.

`PROVISIONING_INTERFACE` - interface to use for ironic, dnsmasq (dhcpd) and httpd.

```bash
sudo docker run -d --net host --privileged \
    --name httpd-infra \
    -v "IMAGES_PATH":/shared \
    --entrypoint /bin/runhttpd \
    --env "PROVISIONING_INTERFACE=ironicendpoint" \
    quay.io/metal3-io/ironic:latest
```

## Ironic in-cluster installation

> For in-of-cluster Ironic installation, we will run a set of containers within

Reviewer 2: ```suggestion
For in-cluster Ironic installation, we will run a set of containers within
```

We assume you are inside the local baremetal-operator path, if not you need to
clone it first and `cd` to the root path.

```bash
 git clone https://github.com/metal3-io/baremetal-operator.git
 cd baremetal-operator
```

Basic authentication enabled:

```bash
 kustomize build ironic-deployment/basic-auth | kubectl apply -f -
```

TLS enabled:

```bash
 kustomize build ironic-deployment/basic-auth/tls | kubectl apply -f -
```

## Ironic out-of-cluster installation

For out-of-cluster Ironic installation, we will run a set of docker containers outside
of a Kubernetes cluster. To pass Ironic settings, you can export corresponding [environmental
variables](#environmental-variables) on the current shell before calling [run_local_ironic.sh](https://github.com/metal3-io/baremetal-operator/blob/main/tools/run_local_ironic.sh)
installation script. This will start below containers:

- ironic
- ironic-inspector
- ironic-endpoint-keepalived
- ironic-log-watch
- ipa-downloader
- dnsmasq
- httpd
- mariadb; if `IRONIC_USE_MARIADB` = "true"

If in-cluster ironic installation, we used different manifests for TLS and basic auth,

> here we are exporting environment variables for enabling/disabling of TLS & basic auth

Reviewer: ```suggestion
here we are exporting environment variables for enabling/disabling TLS & basic auth
```

# Install Ironic

> Baremetal Operator must be able to interact with Ironic. First

Reviewer 3: ```suggestion
Baremetal Operator must be able to interact with Ironic. First, 
```

we will see what are the Ironic optional containers we need to

> run. After we will go through Ironic installation in-cluster and

Reviewer 3: ```suggestion
run. After that we will go through Ironic installation in-cluster and
```

out-of-cluster. In any case, there are seven containers that need

> to run to be able to provision baremetal nodes.

Reviewer 3: ```suggestion
to run to be able in order to provision baremetal nodes.
```

- ironic-dnsmasq
- ironic-api
- ironic-conductor
- ironic-log-watch
- ironic-inspector
- ironic-ipa-downloader
- mariadb

To know more about each container's functionality check the documentation
[here](https://github.com/metal3-io/ironic-image#description).

## Prerequisites

> Container runtime (e.g., docker, podman). Below we use docker.

Reviewer 3: ```suggestion
Container runtime (e.g., docker, podman). Here we use docker.
```

<!-- include these inline so we can style an match variables -->
{{#include ironic_variables.md}}

## Optional containers

In some circumstances, it might be necessary to install some additional containers on
top of Ironic containers. And this might depend on whether your target nodes are
virtual or physical machines, or whether you want to use more stable BMC protocol
like Redfish and etc. Let's check why we would need these optional containers.
We run these optional containers on the host machine, i.e., out-of-cluster regardless
of the Ironic installation method.

### vbmc

If you plan to manage [livbirt](https://en.wikipedia.org/wiki/Libvirt) Virtual Machine (VM),
there is a high chance that you need to install Virtual Baseboard Management
Controller (VBMC), which is virtual implementation of Baseboard Management

> Controller (BMC). For More info on VBMC check official [documentation](https://docs.openstack.org/virtualbmc/latest/).

Reviewer 3: Add a reference link to BMC as well.

Reviewer 3: Would this be good: https://www.ibm.com/docs/en/power9?topic=interfaces-managing-bmc-based-systems

Reviewer 3: The link I provided seems to be quite specific and I could not get a good reference on the matter. 
Would you please ignore my comment and keep that part as is.
Thanks.

> ```bash

Reviewer 3: This bash script needs an intro to tell what it does.

 sudo docker run -d --net host \
    --name vbmc \
     -v "SOME_LOCAL_PATH":/root/.vbmc \
     -v "/root/.ssh":/root/ssh \
     quay.io/metal3-io/vbmc:latest
```

Our VMBC container image is available at [quay.io/metal3-io/vbmc](https://quay.io/repository/metal3-io/vbmc?tag=latest&tab=tags)

### sushy-tools

> Sushy-tools is the same functionality as VBMC tool, except the fact it relies on [Redfish](https://www.dmtf.org/standards/redfish)

Reviewer 3: ```suggestion
Sushy-tools is the same functionality as VBMC tool, except that it relies on [Redfish](https://www.dmtf.org/standards/redfish)
```

> protocol instead of IPMI. For more info, check official [documentation](https://docs.openstack.org/sushy-tools/latest/index.html).

Reviewer 3: ```suggestion
protocol instead of IPMI. For more info, check the official [documentation](https://docs.openstack.org/sushy-tools/latest/index.html).
```

Susy-tool accepts some configuration parameters, which
we pass in the form of a file. Here is an example of `conf.py`
configuration file. To see full list of supported parameters,

> check sushy-tools [documentation](https://docs.openstack.org/sushy-tools/latest/admin/index.html).

Reviewer 3: ```suggestion
check the sushy-tools [documentation](https://docs.openstack.org/sushy-tools/latest/admin/index.html).
```

```python
# The libvirt URI to use. This option enables libvirt drivers.
# qemu+ssh://root@<remote-libvirt-server>/system?keyfile=<path-to-key>
SUSHY_EMULATOR_LIBVIRT_URI = "qemu+ssh://root@192.168.111.1/system?&keyfile=/root/ssh/id_rsa_virt_power&no_verify=1&no_tty=1"
SUSHY_EMULATOR_IGNORE_BOOT_DEVICE = False
SUSHY_EMULATOR_VMEDIA_VERIFY_SSL = False
```

We will mount `conf.py` into the container's `/root/sushy/` path.

```bash
 sudo docker run -d --net host \
    --name sushy-tools \
     -v "SOME_PATH/conf.py":/root/sushy/conf.py \
     -v "/root/.ssh":/root/ssh \
     quay.io/metal3-io/sushy-tools:latest
```

Our sushy-tools container image is available at  [quay.io/metal3-io/sushy-tools](https://quay.io/repository/metal3-io/sushy-tools?tag=latest&tab=tags)

### httpd

We use this container to run Apache web server, which serves our target node images as well as Ironic

> Python Images (IPA) via http for PXE booting. We will mount locally available images container's

Reviewer 3: This needs a bit clarification. 
What content is mounted
Where is it mounted in the container
where is located on the node

Author: thanks, added

`/shared` dir.

> We are not using actual Apache container image here, but instead [quay.io/metal3-io/ironic](https://quay.io/repository/metal3-io/ironic?tag=latest&tab=tags) which has `runhttpd` entry point to start off the httpd.

Reviewer 3: We are actually using Apache, if the base image is httpd/apache.

Author: I can mention that it is based on Apache?

Reviewer 3: That would be good.

# Install Ironic

Baremetal Operator must be able to interact with Ironic. First,
we will see what are the Ironic optional containers we need to
run. After that we will go through Ironic installation in-cluster and
out-of-cluster. In any case, there are seven containers that need
to run in order to provision baremetal nodes.

> - ironic-dnsmasq

Author: I think I took the name from dev-env, but I can see that it has been renamed to dnsmasq in dev-env too. Updating it here too.

- ironic-api

> - ironic-conductor

Reviewer 4: This is outdated

- ironic-log-watch
- ironic-inspector
- ironic-ipa-downloader

> - mariadb

Reviewer 4: missing httpd

To know more about each container's functionality check the documentation

> [here](https://github.com/metal3-io/ironic-image#description).

Reviewer 4: TODO: move this here

## Prerequisites

Container runtime (e.g., docker, podman). Here we use docker.

<!-- include these inline so we can style an match variables -->
{{#include ironic_variables.md}}

## Optional containers

In some circumstances, it might be necessary to install some additional containers on
top of Ironic containers. And this might depend on whether your target nodes are
virtual or physical machines, or whether you want to use more stable BMC protocol

> like Redfish and etc. Let's check why we would need these optional containers.

Reviewer 5: Redfish and IPMI

We run these optional containers on the host machine, i.e., out-of-cluster regardless
of the Ironic installation method.

### vbmc

> If you plan to manage [livbirt](https://en.wikipedia.org/wiki/Libvirt) Virtual Machine (VM),

Reviewer 4: typo: libvirt

> there is a high chance that you need to install Virtual Baseboard Management

Author: I have added a sentence at the beginning.

Controller (VBMC), which is virtual implementation of Baseboard Management
Controller (BMC). For More info on VBMC check official [documentation](https://docs.openstack.org/virtualbmc/latest/).

> To run VMBC container:

Reviewer 5: typo: VBMC

```bash
 sudo docker run -d --net host \
    --name vbmc \
     -v "SOME_LOCAL_PATH":/root/.vbmc \
     -v "/root/.ssh":/root/ssh \
     quay.io/metal3-io/vbmc:latest
```

> Our VMBC container image is available at [quay.io/metal3-io/vbmc](https://quay.io/repository/metal3-io/vbmc?tag=latest&tab=tags)

Reviewer 5: typo: VBMC

### sushy-tools

> Sushy-tools is the same functionality as VBMC tool, except that it relies on [Redfish](https://www.dmtf.org/standards/redfish)

Reviewer 5: s/is/has

> protocol instead of IPMI. For more info, check the official [documentation](https://docs.openstack.org/sushy-tools/latest/index.html).

Reviewer 4: Same comment: we should probably move it to some dev docs rather than what seems a production documentation.

Author: hmm, sure. I will move sushy-tools and vbmc from this doc for now and add a dev sort of guide as follow up

Susy-tool accepts some configuration parameters, which
we pass in the form of a file. Here is an example of `conf.py`
configuration file. To see full list of supported parameters,
check the sushy-tools [documentation](https://docs.openstack.org/sushy-tools/latest/admin/index.html).

```python
# The libvirt URI to use. This option enables libvirt drivers.
# qemu+ssh://root@<remote-libvirt-server>/system?keyfile=<path-to-key>
SUSHY_EMULATOR_LIBVIRT_URI = "qemu+ssh://root@192.168.111.1/system?&keyfile=/root/ssh/id_rsa_virt_power&no_verify=1&no_tty=1"
SUSHY_EMULATOR_IGNORE_BOOT_DEVICE = False
SUSHY_EMULATOR_VMEDIA_VERIFY_SSL = False
```

We will mount `conf.py` into the container's `/root/sushy/` path.

```bash
 sudo docker run -d --net host \
    --name sushy-tools \
     -v "SOME_PATH/conf.py":/root/sushy/conf.py \
     -v "/root/.ssh":/root/ssh \
     quay.io/metal3-io/sushy-tools:latest
```

Our sushy-tools container image is available at  [quay.io/metal3-io/sushy-tools](https://quay.io/repository/metal3-io/sushy-tools?tag=latest&tab=tags)

### httpd

> We use this container to run an Apache web server for hosting node images. Below is the list of images (from `$IMAGES_PATH`) that we mount into the httpd container's `/shared` path.

Reviewer 4: We also use it for TLS when enabled

1. Ironic Python Images (IPA) via http for PXE booting
    - ironic-python-agent.kernel
    - ironic-python-agent.initramfs

2. Target node image in qcow2 format
    - bionic-server-cloudimg-amd64-disk.img

We are not directly using Apache container image here, but instead [quay.io/metal3-io/ironic](https://quay.io/repository/metal3-io/ironic?tag=latest&tab=tags) which has `runhttpd` entry point to start off the httpd.

`PROVISIONING_INTERFACE` - interface to use for ironic, dnsmasq (dhcpd) and httpd.

```bash

> sudo docker run -d --net host --privileged \

Author: One thing I don't like about run-local-ironic script is it has bunch of variables and conditionals which complicates things. For that purpose, I have taken out the actual container run part. But I agree on the point that what's written here might get outdated easily. I will update the document.

I didn't get what kind of deployment file you mean.

> sudo docker run -d --net host --privileged \

Reviewer 2: Can you also clarify in the document why do we run one httpd out of ironic deployment and one inside the deployment ?

    --name httpd-infra \
    -v "IMAGES_PATH":/shared \
    --entrypoint /bin/runhttpd \
    --env "PROVISIONING_INTERFACE=ironicendpoint" \
    quay.io/metal3-io/ironic:latest
```

> ## Ironic in-cluster installation

Reviewer 4: Okay, this is the recommended installation path. I'd put it first and highlight that this is the way we test and support.

Author: by first you mean at the top?

For in-of-cluster Ironic installation, we will run a set of containers within
a single pod in a Kubernetes cluster. You can enable TLS or basic auth or even
disable both for Ironic and inspector communication. Below we will see kustomize
folders that will help us to install Ironic for each mentioned case. In each
of these deployments, a ConfigMap will be created and mounted to the Ironic pod.
The ConfigMap will be populated based on environment variables from
[ironic-deployment/default/ironic_bmo_configmap.env](https://github.com/metal3-io/baremetal-operator/blob/main/ironic-deployment/default/ironic_bmo_configmap.env). As such, update
`ironic_bmo_configmap.env` with your custom values before deploying the Ironic.

We assume you are inside the local baremetal-operator path, if not you need to
clone it first and `cd` to the root path.

```bash
 git clone https://github.com/metal3-io/baremetal-operator.git
 cd baremetal-operator
```

TLS and Basic authentication disabled:

```bash
 kustomize build ironic-deployment/default | kubectl apply -f -
```

> Basic authentication enabled:

Reviewer 4: This requires additional files to be present with credentials.

Author: do you think just mentioning of the need for username & password is enough ?

```bash
 kustomize build ironic-deployment/basic-auth | kubectl apply -f -
```

TLS enabled:

```bash

>  kustomize build ironic-deployment/tls | kubectl apply -f -

Reviewer 4: `ironic-deployment/basic-auth/tls`

```

## Ironic out-of-cluster installation

For out-of-cluster Ironic installation, we will run a set of docker containers outside
of a Kubernetes cluster. To pass Ironic settings, you can export corresponding [environmental
variables](#environmental-variables) on the current shell before calling [run_local_ironic.sh](https://github.com/metal3-io/baremetal-operator/blob/main/tools/run_local_ironic.sh)

> installation script. This will start below containers:

Reviewer 4: Outdated as well

- ironic-dnsmasq
- ironic-api
- ironic-conductor
- ironic-log-watch
- ironic-inspector
- ironic-ipa-downloader
- mariadb

If in-cluster ironic installation, we used different manifests for TLS and basic auth,
here we are exporting environment variables for enabling/disabling of TLS & basic auth
but use the same script.

> TLS and Basic authentication disabled

Reviewer 4: Same

Susy-tool accepts some configuration parameters, which
we pass in the form of a file. Here is an example of `conf.py`
configuration file. To see full list of supported parameters,
check the sushy-tools [documentation](https://docs.openstack.org/sushy-tools/latest/admin/index.html).

```python
# The libvirt URI to use. This option enables libvirt drivers.
# qemu+ssh://root@<remote-libvirt-server>/system?keyfile=<path-to-key>
SUSHY_EMULATOR_LIBVIRT_URI = "qemu+ssh://root@192.168.111.1/system?&keyfile=/root/ssh/id_rsa_virt_power&no_verify=1&no_tty=1"
SUSHY_EMULATOR_IGNORE_BOOT_DEVICE = False
SUSHY_EMULATOR_VMEDIA_VERIFY_SSL = False
```

We will mount `conf.py` into the container's `/root/sushy/` path.

```bash
 sudo docker run -d --net host \
    --name sushy-tools \
     -v "SOME_PATH/conf.py":/root/sushy/conf.py \
     -v "/root/.ssh":/root/ssh \
     quay.io/metal3-io/sushy-tools:latest
```

Our sushy-tools container image is available at  [quay.io/metal3-io/sushy-tools](https://quay.io/repository/metal3-io/sushy-tools?tag=latest&tab=tags)

### httpd

> We use this container to run Apache web server. Below is the list of images (from `$IMAGES_PATH`) that we mount into the httpd

Reviewer 6: ```suggestion
We use this container to run an Apache web server for hosting VM images. Below is the list of images (from `$IMAGES_PATH`) that we mount into the httpd
```

container's `/shared` path.

1. Ironic Python Images (IPA) via http for PXE booting
    - ironic-python-agent.kernel
    - ironic-python-agent.initramfs

2. Target node image in qcow2 format
    - bionic-server-cloudimg-amd64-disk.img

We are not directly using Apache container image here, but instead [quay.io/metal3-io/ironic](https://quay.io/repository/metal3-io/ironic?tag=latest&tab=tags) which has `runhttpd` entry point to start off the httpd.

`PROVISIONING_INTERFACE` - interface to use for ironic, dnsmasq (dhcpd) and httpd.

```bash
sudo docker run -d --net host --privileged \
    --name httpd-infra \
    -v "IMAGES_PATH":/shared \
    --entrypoint /bin/runhttpd \
    --env "PROVISIONING_INTERFACE=ironicendpoint" \
    quay.io/metal3-io/ironic:latest
```

## Ironic in-cluster installation

> For out-of-cluster Ironic installation, we will run set of containers within

Reviewer 6: ```suggestion
For out-of-cluster Ironic installation, we will run a set of containers within
```

a single pod in a Kubernetes cluster. You can enable TLS or basic auth or even
disable both for Ironic and inspector communication. Below we will see kustomize
folders that will help us to install Ironic for each mentioned case. In each
of these deployments, a ConfigMap will be created and mounted to the Ironic pod.

> The ConfigMap will be populated based on environmental variables from

Reviewer 6: ```suggestion
The ConfigMap will be populated based on environment variables from
```

[ironic-deployment/default/ironic_bmo_configmap.env](https://github.com/metal3-io/baremetal-operator/blob/main/ironic-deployment/default/ironic_bmo_configmap.env). As such, update
`ironic_bmo_configmap.env` with your custom values before deploying the Ironic.

We assume you are inside the local baremetal-operator path, if not you need to
clone it first and `cd` to the root path.

```bash
 git clone https://github.com/metal3-io/baremetal-operator.git
 cd baremetal-operator
```

TLS and Basic authentication disabled:

```bash
 kustomize build ironic-deployment/default | kubectl apply -f -
```

Basic authentication enabled:

```bash
 kustomize build ironic-deployment/basic-auth | kubectl apply -f -
```

TLS enabled:

```bash
 kustomize build ironic-deployment/tls | kubectl apply -f -
```

## Ironic out-of-cluster installation

> For out-of-cluster Ironic installation, we will run set of docker containers outside

Reviewer 6: ```suggestion
For out-of-cluster Ironic installation, we will run a set of docker containers outside
```
